app/upgrades/signetlaunch/upgrades.go

```diff
@@ -79,6 +101,115 @@ func LoadBTCHeadersFromData() ([]*btclighttypes.BTCHeaderInfo, error) {
 	return gs.BtcHeaders, nil
 }
 
+// LoadSignedFPsFromData returns the finality providers from the json string.
+// It also verifies if the msg is correctly signed and is valid to be inserted.
+func LoadSignedFPsFromData(cdc codec.Codec, txJSONDecoder sdk.TxDecoder) ([]*btcstktypes.MsgCreateFinalityProvider, error) {
+	buff := bytes.NewBufferString(SignedFPsStr)
+
+	var d dataSignedFps
+	err := json.Unmarshal(buff.Bytes(), &d)
+	if err != nil {
+		return nil, err
+	}
+
+	fps := make([]*btcstktypes.MsgCreateFinalityProvider, len(d.SignedTxsFP))
+	for i, txAny := range d.SignedTxsFP {
+		txBytes, err := json.Marshal(txAny)
+		if err != nil {
+			return nil, err
+		}
+
+		tx, err := txJSONDecoder(txBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		fp, err := parseCreateFPFromSignedTx(cdc, tx)
+		if err != nil {
+			return nil, err
+		}
+
+		fps[i] = fp
+	}
+
+	// sorts all the FPs by their addresses
+	sort.Slice(fps, func(i, j int) bool {
+		return fps[i].Addr > fps[j].Addr
+	})
+
+	return fps, nil
+}
+
+func parseCreateFPFromSignedTx(cdc codec.Codec, tx sdk.Tx) (*btcstktypes.MsgCreateFinalityProvider, error) {
+	msgsV2, err := tx.GetMsgsV2()
+	if err != nil {
+		return nil, err
+	}
+
+	msgs := tx.GetMsgs()
+	if len(msgs) == 0 {
+		return nil, fmt.Errorf("invalid msg, there is no msg inside the tx %+v", tx)
+	}
+
+	// each tx should only contain one msg
+	if len(msgs) != 1 {
+		return nil, fmt.Errorf("each tx should contain only one message, invalid tx %+v", tx)
+	}
+
+	msg, ok := msgs[0].(*btcstktypes.MsgCreateFinalityProvider)
+	if !ok {
+		return nil, fmt.Errorf("unable to parse %+v to MsgCreateFinalityProvider", msg)
+	}
+
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("error validating basic msg: %w", err)
+	}
+
+	msgV2 := msgsV2[0]
+	signers, err := cdc.GetMsgV2Signers(msgV2)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get signers from msg %+v: %w", msg, err)
+	}
+
+	if len(signers) == 0 {
+		return nil, fmt.Errorf("no signer at msg %+v", msgV2)
+	}
```

Comment:
Should we ensure the number of signers is 1?

Reply:
agree

Reply:
done at https://github.com/babylonlabs-io/babylon/pull/28/commits/2bb47c4f305f489af2065402ba56576a1ced60cd